src/runtime/hexagon_remote/remote.cpp

```diff
@@ -44,15 +95,239 @@ int halide_do_task(void *user_context, halide_task_t f, int idx,
+    // ***********************
+    lock_work_queue();
+    FARF(LOW, "HVX_TP: %d: goto_work: got a lock on the work queue\n", tid);
+
+    // If I'm a job owner, then I was the thread that called
+    // do_par_for, and I should only stay in this function while there
+    // are active workers on the job (i.e. running()) or if the job
+    // hasn't been claimed entirely (!claimed()). If I'm a lowly worker
+    // thread, I should stay in this function as long as the work queue is running.
+    while (owned_job != NULL ? (owned_job->running() || !owned_job->claimed())
+           : work_queue.running()) {
+
+        FARF(LOW, "HVX_TP: %d: goto_work: In the main goto_work loop \n", tid);
+        int num_jobs = work_queue.num_jobs, job_idx;
+        work *jobs = work_queue.jobs, *orig_job;
+        bool job_found = false;
+        if (!owned_job) {
+            // If this threads doesn't own a job, it looks for one
+            // and tries to do it. If it cannot find a job, it goes
+            // to sleep.
+            if (num_jobs > 0) {
+                for (job_idx = MAX_WORKLOADS-1 ; job_idx >= 0; job_idx--) {
+                    if (!jobs[job_idx].claimed()) {
+                        myjob = jobs[job_idx];
+                        orig_job = &jobs[job_idx];
+                        job_found = true;
+                        break;
+                    }
+                }
+            }
+            if (!job_found) {
+                // release the lock and go to sleep.
+                // ***************************
+                // *** Work queue unlocked ***
+                // ***************************
+                FARF(LOW, "HVX_TP: %d: goto_work: Couldn't find a job, going to sleep\n", tid);
+                unlock_work_queue();
+                qurt_sem_down(&wait_for_work);
+                FARF(LOW, "HVX_TP: %d: goto_work: Got woken up. Going to look for work.\n", tid);
+                 // ***********************
+                // *** Lock work queue ***
+                // ***********************
+                lock_work_queue();
+                continue;
+            } else {
+                FARF(LOW, "HVX_TP: %d: goto_work: Found a job at idx = %d working on x = %d\n", tid, job_idx, myjob.next);
+            }
+        } else {
+            // We are here only if this thread owns a job that is not done.
+            // So do a part of it.
+            if (owned_job->claimed()) {
+                // the owner goes to sleep till the workers wake it up.
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner about to sleep\n", tid);
+                qurt_cond_wait(&owned_job->wakeup_owner, &work_queue.work_mutex);
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner waking up.\n", tid);
+                // This thread should have the lock now after having been woken up.
+                break;
+            } else {
+                myjob = *owned_job;
+                orig_job = owned_job;
```

Comment:
Whether a thread owns a job or not (i.e regardless of owned_job == NULL), this is the order of business for a thread.
-> lock the work queue.
-> find a job from the work queue.
-> *Mark that you have taken part of the job by incrementing job.next.*
-> *Mark that you are one of the workers on the job by incrementing job.active_workers*
-> Release the lock
-> do the work i.e work on "next". However the next of the job has been incremented thus necessitating a copy of job.
-> Once the work is done, acquire a lock on the work queue again.
-> update exit_status and active_workers--.
-> loop again.

The reason i do orig_job = owned_job is because then I can do the following

`  orig_job->next++;
   orig_job->active_workers++;
`
whether the thread is a worker thread or a master thread.